session/sshforward/sshprovider/agentprovider_windows.go

```diff
@@ -0,0 +1,60 @@
+// +build windows
+
+package sshprovider
+
+import (
+	"net"
+	"regexp"
+	"strings"
+	"syscall"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/pkg/errors"
+)
+
+// Returns the Windows OpenSSH agent named pipe path, but
+// only if the agent is running. Returns an error otherwise.
+func getFallbackAgentPath() (string, error) {
+	// Windows OpenSSH agent uses a named pipe rather
+	// than a UNIX socket. These pipes do not play nice
+	// with os.Stat (which tries to open its target), so
+	// use a FindFirstFile syscall to check for existence.
+	var fd syscall.Win32finddata
+
+	path := `\\.\pipe\openssh-ssh-agent`
```

Comment:
Under what circumstance is a system-wide ssh agent the right source of key material? I'm not overly-familiar with the OpenSSH for Windows client, is the agent maintaining per-user data separation, or is material put here shared across all users?

Reply:
The Windows OpenSSH implementation impersonates the security context of the calling user to access their SSH keys which are stored in the user-specific HKCU registry hive.

See `keyagent-request.c`:
- [process_request_identities()](https://github.com/PowerShell/openssh-portable/blob/46688b7bc07aa556e167543a5935551355b081c8/contrib/win32/win32compat/ssh-agent/keyagent-request.c#L385) reads from
- [get_user_root()](https://github.com/PowerShell/openssh-portable/blob/46688b7bc07aa556e167543a5935551355b081c8/contrib/win32/win32compat/ssh-agent/keyagent-request.c#L54) which opens HKCU hive after using
- [ImpersonateLoggedOnUser](https://docs.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-impersonateloggedonuser) to switch to the calling user's security context.

Hence the use of a system-wide default agent path.

Reply:
Oh, that's very nice. Thank you.